subprojects/scheduler/src/main/java/org/gradle/internal/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,592 @@
+        public void interrupt() {
+            Thread thread = this.thread;
+            if (thread != null) {
+                thread.interrupt();
+            }
+        }
+
+        @Override
+        public void run() {
+            this.thread = Thread.currentThread();
+            WorkerLease workerLease = parentLease.createChild();
+
+            AtomicLong busy = new AtomicLong(0);
+            Timer totalTimer = Time.startTimer();
+            Timer nodeTimer = Time.startTimer();
+
+            while (true) {
+                try {
+                    NodeExecution work = workQueue.take();
+
+                    nodeTimer.reset();
+                    work.runWithLease(workerLease);
+                    long taskDuration = nodeTimer.getElapsedMillis();
+                    busy.addAndGet(taskDuration);
+                    if (LOGGER.isInfoEnabled()) {
+                        LOGGER.info("{} ({}) completed. Took {}.", work, Thread.currentThread(), TimeFormatting.formatDurationVerbose(taskDuration));
+                    }
+
+                    availableWorkers.add(this);
+                } catch (InterruptedException e) {
+                    break;
+                }
+            }
+
+            long total = totalTimer.getElapsedMillis();
+
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Task worker [{}] finished, busy: {}, idle: {}", Thread.currentThread(), TimeFormatting.formatDurationVerbose(busy.get()), TimeFormatting.formatDurationVerbose(total - busy.get()));
+            }
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+    }
+
+    private class NodeExecution {
+        private final Node node;
+        private final NodeExecutor nodeExecutor;
+        private final ResourceLock resourceLock;
+
+        public NodeExecution(Node node, NodeExecutor nodeExecutor, @Nullable ResourceLock resourceLock) {
+            this.node = node;
+            this.nodeExecutor = nodeExecutor;
+            this.resourceLock = resourceLock;
+        }
+
+        public void runWithLease(final WorkerLease workerLease) {
+            boolean acquiredLocks = coordinationService.withStateLock(new Transformer<ResourceLockState.Disposition, ResourceLockState>() {
+                @Override
+                public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
+                    if (!workerLease.tryLock()) {
+                        System.out.printf("<< Failed to secure lease %s for %s%n", workerLease, node);
+                        return FAILED;
+                    }
+                    System.out.printf(">> Acquired lease %s for %s%n", workerLease, node);
+                    if (resourceLock != null && !resourceLock.tryLock()) {
+                        System.out.printf("<< Failed to acquire lock %s for %s, releasing lease %s%n", node, resourceLock, workerLease);
+                        workerLease.unlock();
```

Comment:
No need to unlock here, we return so the coordination service takes care of the unlocking.

Reply:
I would like to build this in a way that it will work even if we remove the coordination-service eventually. It also just feels bad not to unlock things that you've locked.